### DISCUSSION TITLE
Design sensitivity analysis (derivative of results w.r. to parameters)

### DISCUSSION THREAD

**Discussion by WulfHans** [idaholab/moose#18464] [Q&A Modules: Solid mechanics] [answered]
The derivative of results (e.g. displacements) with respect to parameters (e.g. material parameters) is often needed for parameter identification or optimization. Instead of multiple calls to the FE program and using finite differences (costly!), there is an alternative method that calculates these sensitivities directly in the FE code. 
The method boils down to: After computing the convergent solution for the current time-step, you have to solve another system with the same system matrix but a different right hand side. If a decomposition is used for solving, it can be re-used resulting in minimal overhead. The additional right-hand-side contains the derivative of the constitutive law w.r.t. the parameter - this has to be provided by an additional subroutine or computed numerically.

This method has been implemented in several reasearch FE programs under various names like semi-analytical sensitivity analysis. Recently, it has been implemented in ABAQUS as well under the name "Design sensitivity analysis":
https://classes.engineering.wustl.edu/2009/spring/mase5513/abaqus/docs/v6.6/books/stm/default.htm?startat=ch02s18ath57.html

Unfortunately, ABAQUS does not allow any modifications to the process and is very limited in the choice of material laws. Concerning the homegrown FE codes ... I believe it is time to move to something larger and more professional.

The main questions are: **Can this be done with MOOSE with a reasonable effort? Is the framework well-suited for this type of modification? Has someone attempted something similar already?**

The very rough path to implementation looks like:
- define additional field variables for each derivative (should easy via `variable` tag)
- create an additional interface that represents the derivative of the constitutive law w.r.t. the parameters
- compute the RHS using this interface (this might be very similar to existing functions)
- Make sure that the equation system is solved again for each parameter with the new right-hand-sides, after NEWTON converged (my first idea is that you might use a custom class derived from Executioner for this)

Any hints concerning this plan are very welcome as well, especially which are the places to start off and which functionality can be easily reused.

**Comment by GiudGiud** [comment, marked as answer]
Hello

@friedmud has this been done before?
@lindsayad what's your best guess on effort level for this?

I don't see any road blocks in the path to implementation. We can compute variable derivatives. We would initalize a new simulation with the result of the previous one rather than try to do both parts in a single input file and re-use the Jacobian (not really worth the additional complications imo). 
We can compute derivatives of the consistutive laws wrt variables using automatic differentiation, but also can declare derivative material properties and write the derivatives manually for those (if other parameters are involved). 
What do you mean precisely by RHS here? Which terms?

Guillaume

**Reply by WulfHans** [reply]
Thanks for the affirmative replies already!

Concerning the questions:
The RHS in this case is essentially the normal RHS, but derived by the parameter. For the mechanical problem, its nodal forces derived by the parameter (thats why it is often called pseudo-load). It contains two terms, that describe essentially describe 
a) how much does the force directly change due to change in stress and 
b) how much does does it indirectly change by changing the solution
when applying a variation to a parameter.
At the end, the integrals look very similar to the computation of the normal RHS, but with the derivates wrt the parameters instead.
I could write down some equations (is some LaTeX supported here?), but in fact the Abaqus manual linked above nicely describes it as well, without too much overhead.

Btw: If there are some state variables (internal variables of the constitutive law), these need to be stored and retrieved as well. This should be no big problem, right?

**Reply by dschwen** [reply]
If we indeed do this using forward mode automatic differentiation, then all state variables would cary along the derivative w.r.t. the design parameter anyways. I'm not sure there'd even be this special treatment of the RHS. Maybe I'm being overly naive, but in my mind adding the additional parameter derivative entries should make then get carried through the entire solution pretty much automatically.
@lynnmunday can probably comment more (he's the PI on the project).

**Reply by lynnmunday** [reply]
@WulfHans We have implemented a system to solve inverse optimization problems for parameter estimation using the adjoint method to compute the gradient.  We only did this for force inversion and haven't completed it for material identification.  
We are also doing this for discrete measurement points and have not done this for the whole solution field.  Are you only interested in the gradient part of this?  I think that is all you need to compute the sensitivity.

I think @GiudGiud idea on how to compute this sensitivity might work.  To get the gradient at each node, you need something like the patchrecovery auxKernel that can multiply the derivative of the material property with the gradients of the shape functions.  This will need to be done for all of the elements touching the node.  I think I was told a userObject could also do this.  This new nodal value will then need to be multiplied by the nodal solution from the forward problem to get the gradient.  I think this is the right.  

Like @dschwen mentions, we will eventually compute this gradient with automatic differentiation.

**Reply by aaelmeli** [reply]
Hi @lynnmunday,
I have the same problem setting, i.e material identification given discrete responses and utilizing the adjoint formulation to compute the gradient (a derivative of the objective (cost) functional w.r.t material properties). So, if possible, I need some hints on the implementations from you since I think we almost have the same problem. 
summary of the problem:
I am defining the spatial distribution of the material properties through an ASCII file (coordinates and material property value). The spatial material distribution is defined such that each material pixel includes one or more finite elements. 
To compute the gradient vector, each element of the gradient vector will involve computing the derivative of the constitutive equation w.r.t the corresponding material parameter. The computation of the constitutive equation w.r.t that corresponding material parameter would result in a very sparse matrix (nonzero only at the dofs that are located inside that parameter "pixel"). So, I was wondering if you could give some pointers on the implementation part of the gradient, in particular, what functions can be reused from moose, and what kernels/interfaces should be implemented. 
Thank you.

**Reply by lynnmunday** [reply]
@aaelmeli This current pr is for the patchrecovery auxkernel and userobjects:  https://github.com/idaholab/moose/pull/18721#pullrequestreview-743330733
I think we need something like this to compute the gradients.  
To compute this sensitivity at each node you need to be able to compute the following contribitution from each element that touches that node:
nodal_sensitivty = element_integral (grad_shapefunction*grad_forwardSolution)
I don't think nodal aux_kernels have access to element data like shape functions.  That is why @hugary1995 added a userObject in the above pr.  He is computing a nodal flux (i.e. nodal stress) based on the element data which is kind of what you need to do.  So Gary's nodal patch recovery could compute grad_forwardSolution at the node but we need to also also include in this integral the multiplication by the derivative of the shape function.   Maybe this will be easy to do now that Gary has added the patch recovery capability.  I haven't had time to look at it.  I think the sensitivity will be simpler than patchRecovery because Gary is also solving a least squares problem and we don't need that.

**Reply by hugary1995** [reply]
Hey Lynn, two possibly dumb questions: 
- why do we need spatial gradients of material properties?
- why do we need those gradients at the nodes?

**Reply by aaelmeli** [reply]
> Hey Lynn, two possibly dumb questions:
> 
> * why do we need spatial gradients of material properties?
I can reply on this question: in material identification inverse problems (like for example subsurface imaging, biomedical imaging) we are interested in determining the material properties and its spatial distribution. By sending and tracking sound/electromagnetic waves, that inversion was possible. the goal then is to minimize the difference between the measured wavefield and simulated wave field. So, it is optimization problem where the parameters are the material properties and its spatial distribution. Given an objective (cost) functional of your choice, we then need to compute the derivative of that cost functional w.r.t material properties. This requires computing the derivative of the constitutive equation w.r.t the material properties. 
I hope this answers the question.

**Reply by hugary1995** [reply]
Okay I see. Then you want to use nodal path recovery to get the gradient of material property at quadrature points. To be honest, I think that's pretty difficult to implement for general constitutive laws. If it ever gets implemented in moose, I think it will be using what @dschwen proposed below [https://github.com/idaholab/moose/discussions/18464#discussioncomment-1088823](https://github.com/idaholab/moose/discussions/18464#discussioncomment-1088823) .

**Reply by lynnmunday** [reply]
These manual gradients wouldn't be too bad to implement for some simple problems like linear elasticity.  I agree that Daniels approach will be better but I think that might not be ready for a while.  This would be a good check for what Daniel does and would let us work on some other parts of the optimization code while we wait.  

Some more information about the optimization code is available in section 3 of this milestone report:
https://www.osti.gov/biblio/1812088-enhanced-mechanical-property-evaluation-using-innovative-data-analytics-capability

**Comment by dschwen** [comment]
We have an LDRD where this is on our schedule for the beginning of the coming fiscal year.

**Reply by dschwen** [reply]
We basically need the ability to add custom entries to the AD derivative vector. This could be done by using scalar variables to represent the design parameters for example.
